File: agc039_b_bfs.cpp
// https://atcoder.jp/contests/agc039/tasks/agc039_b
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
#include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

int n;
vector<vector<int>> to;

int bfs(int sv) {
    queue<int> q;
    vector<int> dist(n, INF);
    q.push(sv);
    dist[sv] = 0;
    int res = 0;
    while (q.size()) {
        int v = q.front(); q.pop();
        res = max(res, dist[v]+1);
        for (int u: to[v]) {
            if (dist[u] != INF) continue;
            dist[u] = dist[v] + 1;
            q.push(u);
        }
    }
    rep(i,n) {      // 2部グラフ判定
        for (int j : to[i]) {
            if (abs(dist[i]-dist[j]) != 1) return -1;
        }
    }
    return res;
}

int main() {
    cin >> n;
    to = vector<vector<int>>(n);
    rep(i,n) {
        string s;
        cin >> s;
        rep(j,n) {
            if(s[j]=='1') to[i].push_back(j);
        }
    }
    int ans = -1;
    rep(i,n) ans = max(ans, bfs(i));
	cout << ans << endl;
	return 0;
}

File: atcoder/agc039_b_bfs.py
# https://atcoder.jp/contests/agc039/tasks/agc039_b
import sys
# input = sys.stdin.buffer.readline
def input(): return sys.stdin.readline().rstrip()
sys.setrecursionlimit(10 ** 7)

from collections import deque

N = int(input())
S = [input() for _ in range(N)]

def bfs(start):
    res = 0
    dist = [-1] * N
    dist[start] = 1
    q = deque()
    q.append([start, 1])
    st = set()
    st.add((start, 1))

    while q:
        i, d = q.popleft()
        if dist[i] == -1:
            dist[i] = d
            res = max(res, d)
        elif dist[i] != d:
            res = -1
            break
        for j in range(N):
            if S[i][j] == '1' and dist[j] != d-1:
                if (j, d+1) not in st:
                    q.append([j, d+1])
                    st.add((j, d+1))
    return res

ans = -1
for i in range(N):
    k = bfs(i)
    if k == -1:
        break
    else:
        ans = max(ans, k)

print(ans)


# TLE
# import sys
# # input = sys.stdin.buffer.readline
# def input(): return sys.stdin.readline().rstrip()
# sys.setrecursionlimit(10 ** 7)

# INF = 1001001001
# from scipy.sparse.csgraph import floyd_warshall
# from collections import deque

# N = int(input())
# S = [input() for _ in range(N)]
# G = [[INF]*N for _ in range(N)]
# for i in range(N):
#     for j in range(N):
#         if i == j:
#             G[i][j] = 0
#         elif S[i][j] == '1':
#             G[i][j] = 1
# # print(G)

# dist = floyd_warshall(G)
# # print(dist)

# k_1 = 0
# for i in range(N):
#     for j in range(N):
#         if int(dist[i][j]) > k_1:
#             k_1 = int(dist[i][j])
#             max_i = i
#             max_j = j
# k = k_1 + 1

# dist = [-1] * N
# dist[max_i] = 1
# q = deque()
# q.append([max_i, 1])

# while q:
#     i, d = q.popleft()
#     if dist[i] == -1:
#         dist[i] = d
#     elif dist[i] != d:
#         k = -1
#         break
#     for j in range(N):
#         if S[i][j] == '1' and dist[j] != d-1:
#             q.append([j, d+1])

# print(k)
